Lib/test/test_concurrent_futures/test_deadlock.py

```diff
@@ -241,6 +243,65 @@ def test_crash_big_data(self):
 
         executor.shutdown(wait=True)
 
+    def test_gh105829_should_not_deadlock_if_wakeup_pipe_full(self):
+        # Issue #105829: The _ExecutorManagerThread wakeup pipe could
+        # fill up and block. See: https://github.com/python/cpython/issues/105829
+
+        # Lots of cargo culting while writing this test, apologies if
+        # something is really stupid...
+
+        self.executor.shutdown(wait=True)
+
+        if not hasattr(signal, 'alarm'):
+            raise unittest.SkipTest(
+                "Tested platform does not support the alarm signal")
+
+        def timeout(_signum, _frame):
+            raise RuntimeError("timed out while submitting jobs?")
+
+        thread_run = futures.process._ExecutorManagerThread.run
+        def mock_run(self):
+            # Delay thread startup so the wakeup pipe can fill up and block
+            time.sleep(5)
+            thread_run(self)
+
+        def adjust_and_check_jobs_needed_to_block_pipe(connection):
+            try:
+                # Try to reduce pipe size to speed up test. Only works on Unix systems
+                import fcntl
+                from fcntl import F_SETPIPE_SZ
+                pipe_size = fcntl.fcntl(connection.fileno(), F_SETPIPE_SZ, 1024)
+            except ImportError:
+                # Assume 64k pipe if we fail, makes test take longer
+                pipe_size = 65536
+
+            # We send 4 bytes per job (one zero sized bytes object)
+            return pipe_size // 4 + 100  # Add some margin
+
+        with unittest.mock.patch.object(futures.process._ExecutorManagerThread, 'run', mock_run):
+            with self.executor_type(max_workers=2,
+                                    mp_context=self.get_context()) as executor:
+                self.executor = executor  # Allow clean up in fail_on_deadlock
+
+                # Try to speed up the test by reducing the size of the wakeup pipe
+                job_num = adjust_and_check_jobs_needed_to_block_pipe(
+                    executor._executor_manager_thread_wakeup._writer)
+                job_data = range(job_num)
+
+                # Need to use sigalarm for timeout detection because
+                # Executor.submit is not guarded by any timeout (both
+                # self._work_ids.put(self._queue_count) and
+                # self._executor_manager_thread_wakeup.wakeup() might
+                # timeout, maybe more?). In this specific case it was
+                # the wakeup call that deadlocked on a blocking pipe.
+                old_handler = signal.signal(signal.SIGALRM, timeout)
+                try:
+                    signal.alarm(int(support.LONG_TIMEOUT))
+                    self.assertEqual(job_num, len(list(executor.map(int, job_data))))
+                finally:
+                    signal.alarm(0)
+                    signal.signal(signal.SIGALRM, old_handler)
```

Comment:
@tomMoral - would you be able to review this test? Any ways we could add more testing around this or make this test quicker would be great :-)

Reply:
```python
    def wakeup(self):
        if not self._closed and not self._set:
            self._set = True
            self._writer.send_bytes(b"")

    def clear(self):
        if not self._closed:
            self._set = False
            while self._reader.poll():
                self._reader.recv_bytes()
```

Is there any way to get this to work without adding back a lock between `wakeup()` and `clear()`? I don't think so? We need to guard against both the data race with self._set and make sure the changes to self._set and the access to the reader/writer is atomic I think?

Reply:
I think this is similar to the original patch where we made the pipe non-blocking and ignored the write if it blocked. This worked because wakeup and clear where both guarded by the shutdown lock

Reply:
I think I'm in favour of keeping the fix currently in this PR, I believe both myself and @sharvil have been testing this with real world use for a couple of weeks now.

If we can get the test faster but still a reliable catch, then let's do that, but otherwise I'm keen to land this to get into the soonest possible 3.11 bugfix release :-)

Reply:
Yes I think this woudl work without a lock if actually, we set:

```python
    def clear(self):
        if not self._closed:
            while self._reader.poll():
                self._reader.recv_bytes()
            self._set = False
```
This way, there should be no way to have a situation where `self._set` is `True` with no message in the pipe.

But I agree this might be a bit more error prone, so if we want to move fast, keeping the proposed patch seems sufficient.
For faster tests, we could mock a blocking `_ThreadWakeup` object with a very small queue size with something that block in python instead of really going for the blocking behavior of the pipe?

Reply:
something like this?

```python
        class MockWakeup(_ThreadWakeup):
            def __init__(self):
                super().__init__()
                self._dummy_queue = queue.Queue(maxsize=1)

            def wakeup(self):
                self._dummy_queue.put(None, block=True)
                super().wakeup()

            def clear(self):
                try:
                    while True:
                        self._dummy_queue.get_nowait()
                except queue.Empty:
                    pass
                super().clear()
```

Reply:
I pushed an update to the test based on this strategy